### ISSUE TITLE
RLT online RL fine-tuning for XVLA in Gazebo sim

### ISSUE THREAD

**Issue by vincewu51** [sl628/aic#43]
# RLT Online RL Fine-Tuning for XVLA

## Problem

Behavioral cloning (BC) learns from static demonstrations but never receives feedback about task success. The model cannot discover better strategies than what the expert demonstrated, and cannot learn from its failures during execution. RL fine-tuning addresses this by optimizing a task reward signal through online interaction with the sim.

The RLT (Reinforcement Learning for Transformers) package already implements:
- **RL Token**: A lightweight encoder-decoder that compresses VLA embeddings into a 2048-dim latent `z_rl`
- **Actor-Critic**: TD3-style twin-Q critic + Gaussian actor that outputs action chunks from `z_rl`
- **Offline training**: Pretrain RL token → train actor-critic from demonstration replay buffer
- **Online training loop**: Architecture exists at `trainer.py:1012` but **environment wrapper is a stub**

The missing piece is connecting RLT's online training loop to the actual AIC Gazebo sim.

---

## Architecture Overview

```
┌─────────────────────────────────────────────────────────────────┐
│                     RLT Online Training Loop                     │
│                                                                  │
│  ┌──────────┐    ┌──────────────┐    ┌──────────────────────┐   │
│  │ Env      │───▶│ RL Trainer   │───▶│ Actor (policy)       │   │
│  │ step()   │    │ collect      │    │ predicts z_rl → act  │   │
│  │ → obs,r  │    │ transition   │    └──────────┬───────────┘   │
│  └──────────┘    └──────┬───────┘               │               │
│                         │                       │               │
│                    ┌────▼───────┐         ┌──────▼───────────┐  │
│                    │ Replay     │         │ VLA Backend      │  │
│                    │ Buffer     │         │ extracts z_rl    │  │
│                    └────────────┘         │ from obs         │  │
│                                           └──────────────────┘  │
└─────────────────────────────────────────────────────────────────┘
```

---

### Phase 1: Real AICEnvWrapper (3-4 hours)

**File:** `aic_utils/aic_rlt/scripts/train.py`, lines 148-186 (replace `AICEnvWrapper`)

The current stub:
```python
class AICEnvWrapper:
    def reset(self):
        return {"dummy": True}        # ✗ stub
    def step(self, action_chunk):
        return ({"dummy": True}, 0.0, False, {})  # ✗ stub
```

**New implementation** that communicates with the Gazebo sim:

```python
class AICEnvWrapper:
    def __init__(self, vla, prop_dim=26, sim_config=None):
        self.vla = vla
        self.prop_dim = prop_dim
        # ROS 2 node for sim communication
        self.node = rclpy.create_node("rlt_train_env")
        self.cmd_pub = self.node.create_publisher(MotionUpdate, "/motion_command", 10)
        self.obs_sub = self.node.create_subscription(Observation, "/observation", self._obs_cb, 10)
        self._latest_obs = None
        # Reward config
        self.reward_fn = RewardConfig()  # from trainer.py

    def reset(self):
        # Request sim reset via ROS service
        # Wait for first observation
        return self._wait_for_obs()

    def step(self, action_chunk):
        # action_chunk: (C, 7) numpy array
        rewards = []
        for i in range(len(action_chunk)):
            self._send_action(action_chunk[i])
            obs = self._wait_for_obs()
            reward = self._compute_reward(obs)
            rewards.append(reward)
        # Return last obs + cumulative reward
        return obs, sum(rewards), self._check_done(obs), {}
```

This requires:
1. Importing `rclpy` and ROS 2 message types
2. A ROS 2 node that runs in a separate thread (non-blocking for the training loop)
3. Reward computation from `trainer.py`'s `RewardConfig._compute_structured_reward()`
4. Sim reset via ROS service call to the AIC engine

**Alternative (simpler) approach**: Instead of building a full ROS wrapper, reuse the Docker eval pipeline. Create a script that:
1. Launches the eval Docker stack with `ground_truth:=false` (policy mode)
2. Exposes the policy endpoint as a local HTTP API
3. The env wrapper sends actions via HTTP and receives observations via the same API
4. This avoids the complexity of ROS 2 integration in the training process

### Phase 2: Reward Integration (1-2 hours)

**File:** `aic_utils/aic_rlt/aic_rlt/trainer.py`, lines 258-316 (reward computation)

The `_compute_structured_reward()` method already computes multi-term rewards from the 26D state + goal pose. For online training, this needs to be called from the env wrapper's `step()`.

Key reward terms:
- **Position Gaussian**: `exp(-||pos - goal||² / σ²)` — dense shaping towards port
- **Orientation alignment**: `exp(-θ² / σ_θ²)` — align gripper with port axis
- **Insertion depth**: ramp from 0 to 1 as TCP moves along the port axis
- **Contact bonus**: activates when `axial_tcp_err` exceeds a threshold (indicates contact with port)
- **Success bonus**: large reward when insertion is detected (via `plug_port_dist`)

For online training, **no goal pose is available** (unlike offline where the goal is the last frame of the episode). We need one of:
1. A fixed goal pose from the world frame (port location is known in sim)
2. An automatically inferred goal from the current task description
3. The port pose from TF (available via `ground_truth:=true` during training)

**Recommendation**: Use the ground-truth port pose from TF for training (option 3). This is available in the sim during evaluation and can be exposed via an additional ROS topic.

### Phase 3: Online Training Loop Integration (2-3 hours)

**File:** `aic_utils/aic_rlt/aic_rlt/trainer.py`, lines 1012-1057 (`train()` method)

The `train()` method:
```python
def train(self, initial_obs_fn: Callable) -> None:
```

It calls `_collect_transition()` (line 847) in a loop, which in turn calls `env_step(actions)` (a callback set at init). The method expects:
1. `initial_obs_fn():` Resets env, returns first observation
2. `env_step(actions):` Executes a chunk, returns `(next_obs, reward, done, info)`
3. `get_vla_embeddings(obs):` Extracts VLA internal embeddings from observation
4. `get_vla_action_chunk(obs):` Gets VLA reference action for the observation
5. `human_intervention():` Optional override

The integration points are:
- `initial_obs_fn` = `AICEnvWrapper.reset()`
- `env_step` = `AICEnvWrapper.step()`
- `get_vla_embeddings` = `VLABackend.encode_obs()` (already implemented in `xvla_backend.py`)
- `get_vla_action_chunk` = `VLABackend.get_action()` (already implemented)
- `human_intervention` = stub returning None (DAgger integration for later)

### Phase 4: Launch & Orchestration (2 hours)

Create `scripts/rlt_online_train.sh`:

```bash
#!/usr/bin/env bash
# RLT online RL fine-tuning in AIC sim

# 1. Launch Docker eval stack in training mode
bash scripts/eval_docker.sh <model-id> --build --pull

# 2. Start RLT training with env wrapper connected to sim
python3 -m aic_rlt.scripts.train \
    --mode online_rl \
    --checkpoint_dir ~/aic_rlt_checkpoints/v15_online/ \
    --vla_model_dir /home/yifeng/models/xvla-base \
    --phase2_checkpoint ~/aic_rlt_checkpoints/v15_offline/phase2_offline.pt \
    --config online_rl_config.yaml
```

Config file `online_rl_config.yaml`:
```yaml
reward:
  w_pos: 1.0
  w_ori: 0.3
  w_depth: 0.5
  w_contact: 0.2
  w_success: 10.0
  sigma_pos: 0.03
training:
  total_timesteps: 100000
  batch_size: 256
  learning_rate: 3e-4
  eval_interval: 5000
env:
  task_timeout_s: 180
  control_period_s: 0.25
  sim_docker_compose: ~/aic/docker/compose.yaml
```

---

## File Changes Summary

| File | Change |
|------|--------|
| `aic_utils/aic_rlt/scripts/train.py` | Replace `AICEnvWrapper` stub with real Gazebo environment interface |
| `aic_utils/aic_rlt/aic_rlt/trainer.py` | Expose `RewardConfig._compute_structured_reward()` for online use; expose port pose via callback or config |
| `aic_utils/aic_rlt/aic_rlt/vla/xvla_backend.py` | Add method to extract goal pose from observation or task info |
| `scripts/rlt_online_train.sh` | New: orchestrate online RL training |
| `scripts/rlt_online_config.yaml` | New: RLT online training configuration |
| `docker/aic_xvla/entrypoint_single.sh` | Add env var for training mode (longer timeout, no score limits per trial) |

---

## Prerequisites

Before starting online RL, the offline phases must be complete:

- [ ] Phase 1: RL Token pretrained (`pretrain_rl_token` mode) — produces encoder-decoder weights
- [ ] Phase 2 offline: Actor-critic trained from demonstrations (`offline_rl` mode) — produces initial policy
- [ ] VLA embeddings prepared for training dataset (`prepare_embeddings.py`)

The online phase then fine-tunes the actor (and optionally the RL token encoder) through sim interaction.

---

## Verification

1. **Warm-start test**: Run RL training without updating weights (frozen actor). Verify the sim loops correctly and rewards are computed.
2. **Gradient test**: Run 100 training steps, verify loss decreases and Q-values increase.
3. **Short RL run**: 10K timesteps (~30 min). Check if insertion success rate improves from 0%.
4. **Full run**: 100K timesteps (~5 hours). Compare final sim score against BC baseline.

---

## Estimated Effort

| Phase | Time |
|-------|:----:|
| Phase 1: Real AICEnvWrapper | 3-4 hours |
| Phase 2: Reward integration | 1-2 hours |
| Phase 3: Online loop integration | 2-3 hours |
| Phase 4: Launch & orchestration | 2 hours |
| **Total** | **8-11 hours** |

---

## Risks

1. **Sim stability**: Long-running RL training may cause Gazebo instability or memory leaks. Mitigation: restart Docker stack every N episodes.
2. **Reward hacking**: The model may find unintended ways to maximize the reward (e.g., spinning the gripper). Mitigation: carefully design reward terms, add smoothness penalties.
3. **Sample efficiency**: 100K timesteps at 4Hz control = ~7 hours of sim time. The RL token approach is designed to be sample-efficient (tuning only a 2048-dim latent + MLP actor), but may still require many episodes. Mitigation: start with a well-trained BC policy (Phase 2 offline) before online fine-tuning.
4. **Safety**: The actor may explore actions that damage the sim (e.g., driving the arm into singularities). Mitigation: action limits, velocity clamping, and manual override via `human_intervention()`.
5. **Checkpoint management**: RL training produces many intermediate checkpoints. Implement automatic saving of best-performing checkpoints based on evaluation episodes.
